src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,214 @@
+            {
+                return;
+            }
+
+            var sourceContainer = TextView.TextBuffer.AsTextContainer();
+            if (sourceContainer is null)
+            {
+                return;
+            }
+
+            if (!Workspace.TryGetWorkspace(sourceContainer, out var workspace))
+            {
+                return;
+            }
+
+            var document = sourceContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                return;
+            }
+
+            var option = workspace.Options.GetOption(InlineDiagnosticsOptions.Location, document.Project.Language);
+            if (option == InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+            {
+                var normalizedCollectionSpan = new NormalizedSnapshotSpanCollection(TextView.TextViewLines.FormattedSpan);
+                UpdateSpans_CallOnlyOnUIThread(normalizedCollectionSpan, removeOldTags: true);
+            }
+        }
+
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            if (AdornmentLayer is not null)
+            {
+                var elements = AdornmentLayer.Elements;
+                foreach (var element in elements)
+                {
+                    var tag = (InlineDiagnosticsTag)element.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var format = GetFormat(classificationType);
+                    tag.UpdateColor(format, element.Adornment);
+                }
+            }
+        }
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that it can only display the first one that appears on the line
+        /// </summary>
+        private IDictionary<int, (IMappingTagSpan<InlineDiagnosticsTag> mapTagSpan, SnapshotSpan snapshotSpan)> GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            if (changedSpanCollection.IsEmpty())
+            {
+                return SpecializedCollections.EmptyDictionary<int, (IMappingTagSpan<InlineDiagnosticsTag>, SnapshotSpan)>();
+            }
+
+            var map = new Dictionary<int, (IMappingTagSpan<InlineDiagnosticsTag> mapTagSpan, SnapshotSpan snapshotSpan)>();
+            var viewLines = TextView.TextViewLines;
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
```

Comment:
so question on how this works.  if this is filtering only to TextViewLines, what ensure that as we scroll around we then recompute for what moves into view?  does that happen elsewhere?  (if so, great.  i just want to understand for my mental model).

Reply:
The abstract adornment manager handles the OnLayoutChanged event, which adds the tags when scrolling